cpp/src/parquet/encoding_test.cc

```diff
@@ -862,5 +862,208 @@ TEST_F(DictEncoding, CheckDecodeIndicesNoNulls) {
+
+  const uint32_t value_as_uint = *reinterpret_cast<const uint32_t*>(&value);
+  ASSERT_EQ(static_cast<uint8_t>(value_as_uint & 0xFFU), mutable_data[0]);
+  ASSERT_EQ(static_cast<uint8_t>((value_as_uint >> 8U) & 0xFFU), mutable_data[1]);
+  ASSERT_EQ(static_cast<uint8_t>((value_as_uint >> 16U) & 0xFFU), mutable_data[2]);
+  ASSERT_EQ(static_cast<uint8_t>((value_as_uint >> 24U) & 0xFFU), mutable_data[3]);
+
+  ASSERT_EQ(0, encoder->EstimatedDataEncodedSize());
+
+  encoded_buffer = encoder->FlushValues();
+  ASSERT_EQ(0, encoded_buffer->size());
+}
+
+// Check that the encoder can handle arbitrary large input.
+TEST(ByteStreamSplitEncodeDecode, EncodeLargeInput) {
+  const size_t nvalues = 10000U;
+  std::vector<float> draws(nvalues);
+  GenerateData<float>(nvalues, draws.data(), NULL);
+
+  std::unique_ptr<TypedEncoder<FloatType>> encoder =
+      MakeTypedEncoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  encoder->Put(draws.data(), static_cast<int>(draws.size()));
+  std::shared_ptr<Buffer> encoded_buffer = encoder->FlushValues();
+  ASSERT_EQ(draws.size() * sizeof(float), encoded_buffer->size());
+
+  size_t byte_index = 0U;
+  const uint8_t* encoded_buffer_raw = encoded_buffer->mutable_data();
+  for (size_t i = 0U; i < sizeof(float); ++i) {
+    for (size_t j = 0U; j < draws.size(); ++j) {
+      const float value = draws[j];
+      const uint32_t value_as_uint = *reinterpret_cast<const uint32_t*>(&value);
+      const uint8_t byte =
+          static_cast<const uint8_t>((value_as_uint >> (8U * i)) & 0xFFU);
+      ASSERT_EQ(byte, encoded_buffer_raw[byte_index]);
+      ++byte_index;
+    }
+  }
+}
+
+// Check that the decoder can handle empty input.
+TEST(ByteStreamSplitEncodeDecode, DecodeZeroLenInput) {
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  decoder->SetData(0, NULL, 0);
+  ASSERT_EQ(0U, decoder->Decode(NULL, 0));
+}
+
+TEST(ByteStreamSplitEncodeDecode, DecodeOneLenInput) {
+  std::unique_ptr<TypedDecoder<FloatType>> decoder =
+      MakeTypedDecoder<FloatType>(Encoding::BYTE_STREAM_SPLIT);
+  const uint8_t data[] = {0xDEU, 0xC0U, 0x37U, 0x13U};
+  decoder->SetData(1, data, 4);
+
+  float value = 0U;
+  const int num_decoded = decoder->Decode(&value, 1);
+  ASSERT_EQ(1, num_decoded);
+
+  const uint32_t value_as_uint = *reinterpret_cast<const uint32_t*>(&value);
+  ASSERT_EQ(0x1337C0DEU, value_as_uint);
+}
+
+// Check that requesting to decode more elements than is available in the storage
+// of the decoder works correctly.
+TEST(ByteStreamSplitEncodeDecode, DecodeLargerPortion) {
+  std::unique_ptr<TypedDecoder<DoubleType>> decoder =
+      MakeTypedDecoder<DoubleType>(Encoding::BYTE_STREAM_SPLIT);
+  const uint8_t data[] = {0xDEU, 0xC0U, 0x37U, 0x13U, 0x11U, 0x22U, 0x33U, 0x44U,
+                          0xAAU, 0xBBU, 0xCCU, 0xDDU, 0x55U, 0x66U, 0x77U, 0x88U};
+  decoder->SetData(2, data, 8);
```

Comment:
Shouldn't this be `SetData(2, data, 16)`? (2 double values -> 16 bytes len)

Reply:
This is me trying to test having number of elements greater than what can be supported with the number of bytes provided. This should be in fact a failing test but the implementation doesn't check for this and also the test doesn't check that results are incorrect.
I suppose since the decoder/encoder is only used internally the correct arguments should always be provided, so I shouldn't be testing for that.